### PR TITLE
[WIP] feat: add better support for chrome.Tab and chrome.tabs.query

### PR DESCRIPTION
**Pull request by MarshallOfSound**
#### Description of Change
* Adds supports for most of the properties of the `Tab` class.
* Implements `chrome.tabs.query`

TODO: Tests?

#### Release Notes

Notes: Improved `chrome.tabs` extension API support

### CODE REVIEW

File: lib/renderer/chrome-api.ts [hunk 1]
```diff
@@ -5,8 +5,30 @@ import * as url from 'url'
...
+const ProxyTabProperty = <G extends keyof Electron.WebContents>(getMethod: G, mapGet?: (val: any) => any) => (target: Tab, propertyKey: keyof chrome.tabs.Tab) => {
+  Object.defineProperty(target, propertyKey, {
+    get: function (this: Tab): any {
+      const val = ipcRendererUtils.invokeSync('CHROME_TABS_PROPERTY_GET', this.id, getMethod)
+      return mapGet ? mapGet(val) : val
+    },
+    configurable: false,
+    enumerable: true
+  })
+}
```
Comment (samuelmaddock) [inline lib/renderer/chrome-api.ts]: so good 💯

File: lib/renderer/chrome-api.ts [hunk 2]
```diff
@@ -5,8 +5,30 @@ import * as url from 'url'
...
+  public index = 0
+  public highlighted = false
+  @ProxyTabProperty('isFocused') public active!: boolean
+  public pinned = false
+  @ProxyTabProperty('isCurrentlyAudible') public audible!: boolean
+  @ProxyTabProperty('isDestroyed') public discarded!: boolean
+  public autoDiscardable = false
+  @ProxyTabProperty('getURL') public url!: string
+  @ProxyTabProperty('getTitle') public title!: string
+  @ProxyTabProperty('isLoadingMainFrame', (loading: boolean) => loading ? 'loading' : 'complete') public status!: string
```
Comment (samuelmaddock) [inline lib/renderer/chrome-api.ts]: What's requiring the `!` operator in these cases?

File: lib/renderer/chrome-api.ts [hunk 3]
```diff
@@ -5,8 +5,30 @@ import * as url from 'url'
...
+  public incognito = false
```
Comment (samuelmaddock) [inline lib/renderer/chrome-api.ts]: Would it be worthwhile to implement the rest of the [`Tab`](https://developer.chrome.com/extensions/tabs#type-Tab) properties with placeholders? `windowId`, `width`, `height`, etc.

File: lib/renderer/chrome-api.ts [hunk 4]
```diff
@@ -5,8 +5,30 @@ import * as url from 'url'
...
+      return mapGet ? mapGet(val) : val
+    },
+    configurable: false,
+    enumerable: true
+  })
+}
+
 class Tab {
   public id: number
+  public index = 0
```
Comment (samuelmaddock) [inline lib/renderer/chrome-api.ts]: Could we make this class implement `chrome.tabs.Tab`?

File: lib/renderer/chrome-api.ts [hunk 5]
```diff
@@ -173,6 +195,16 @@ export function injectTo (extensionId: string, isBackgroundPage: boolean, contex
   }
 
   chrome.tabs = {
+    query (
```
Comment (samuelmaddock) [inline lib/renderer/chrome-api.ts]: Could reuse types here
```ts
query: typeof chrome.tabs.query = (info, tabCallback) => ...
```

File: package.json [hunk 6]
```diff
@@ -5,6 +5,7 @@
   "description": "Build cross platform desktop apps with JavaScript, HTML, and CSS",
   "devDependencies": {
     "@octokit/rest": "^16.3.2",
+    "@types/chrome": "0.0.81",
```
Comment (samuelmaddock) [inline package.json]: nice

**Comment by MarshallOfSound** [comment]
Closing out in favor of @samuelmaddock 's extension work

**Pull request closed without merging**